fms/models/roberta.py

```diff
@@ -0,0 +1,321 @@
+    nlayers: int = 12
+    pad_id: int = 1
+    hidden_grow_factor: float = 4.0
+    activation_fn: str = "gelu"
+    classifier_activation_fn: str = "tanh"
+    max_pos: int = 512
+    p_dropout: float = 0.1
+    multiquery_attn: bool = False
+    norm_eps: float = 1e-12
+    tie_heads: bool = False
+
+
+class RoBERTaBlock(nn.Module):
+    def __init__(self, config: RoBERTaConfig):
+        super().__init__()
+        self.config = config
+
+        self.ln = nn.LayerNorm(self.config.emb_dim, self.config.norm_eps)
+        self.ff_ln = nn.LayerNorm(self.config.emb_dim, self.config.norm_eps)
+
+        self.attn = MultiHeadAttention(
+            self.config.emb_dim,
+            self.config.emb_dim // self.config.nheads,
+            self.config.emb_dim // self.config.nheads,
+            self.config.nheads,
+            kvheads=1 if self.config.multiquery_attn else self.config.nheads,
+            p_dropout=self.config.p_dropout,
+            use_bias=True,
+        )
+
+        self.ff_sub_layer = FeedForwardBlock(
+            self.config.emb_dim,
+            hidden_grow_factor=self.config.hidden_grow_factor,
+            activation_fn=str_to_activation(self.config.activation_fn),
+            p_dropout=self.config.p_dropout,
+            use_bias=True,
+        )
+
+        if self.config.p_dropout != 0:
+            self.dropout = nn.Dropout(self.config.p_dropout)
+
+    def forward(
+        self,
+        x: torch.LongTensor,
+        *,
+        mask: Optional[torch.Tensor] = None,
+        attn_algorithm: Optional[str] = None,
+    ):
+        # first we do MHA
+        residual = x
+        # self attention
+        x = self.attn(
+            q=x,
+            k=x,
+            v=x,
+            mask=mask,
+            attn_algorithm=attn_algorithm,
+            is_self=True,
+            is_causal_mask=False,
+        )
+
+        if self.config.p_dropout != 0:
+            x = self.dropout(x)
+
+        # residual connection
+        x = x + residual
+        # post ln
+        x = self.ln(x)
+
+        # then we do FF and Add&Norm
+        residual = x
+        x = self.ff_sub_layer(x)
+
+        if self.config.p_dropout != 0:
+            x = self.dropout(x)
+
+        # another residual
+        x = x + residual
+        x = self.ff_ln(x)
+
+        return x
+
+
+class RoBERTaHeadless(nn.Module):
+    def __init__(
+        self, config: RoBERTaConfig, distributed_strategy: DistributedStrategy
+    ):
+        super().__init__()
+        self.config = config
+
+        self.layers = nn.ModuleList(
+            [
+                distributed_strategy.distribute_layer(RoBERTaBlock(self.config), i)
+                for i in range(self.config.nlayers)
+            ]
+        )
+
+        self.embedding = nn.Embedding(self.config.src_vocab_size, self.config.emb_dim)
+
+        self.position_embedding = nn.Embedding(self.config.max_pos, self.config.emb_dim)
+
+        self.enc_norm = distributed_strategy.distribute_module(
```

Comment:
have we done any kind of test to see if this distributed code works properly on this model? we can probably load an instance, apply_tp(roberta_model) with multiple ranks and save a sharded checkpoint, then try to load the tp-sharded checkpoint via get_models(.... , distributed_strategy='tp') which would hit the relevant distribution code.

Reply:
I tested that I was able to save and load a tp sharded checkpoint as well as run the forward pass with the same results before and after. I decided to not tp-shard the embedding for RoBERTa as well as the head as it would overly complicate the code, as well as make the model much less useable as the vocab size in roberta was not built with tp in mind. (I was able to get tp working, however given the constraints, I do not think keeping it for the embedding/head is the best solution)